### PR TITLE
Nll masking

### PR DESCRIPTION
**Pull request by DeNeutoy**
This follows on from the SRL model PR, implementing masked cross entropy loss for the sequence models we have. The test is a bit rubbish - any suggestion for a better one would be appreciated.

### CODE REVIEW

**Comment by DeNeutoy** [comment]
Bah, I screwed up the merge/rebase on this. I thought if I rebase against another branch, the commits aren't copied?

**Comment by matt-gardner** [comment]
Should I review this now, or are you going to fix the commits still?

**Comment by DeNeutoy** [comment]
@matt-gardner sorry the rebase took longer than I thought because I kept introducing new merge conflicts every time I fixed the previous one.

**Comment by matt-gardner** [comment]
No problem, I just wasn't sure if you were ready or not.

**Comment by matt-gardner** [comment]
Oh, one more request - I just merged the BiDAF PR - would you mind switching the loss function for that model, too?

File: allennlp/common/tensor.py [hunk 1]
```diff
@@ -94,6 +94,20 @@ def arrays_to_variables(data_structure: Dict[str, Union[dict, numpy.ndarray]],
             return torch_variable.cuda(cuda_device)
 
 
+def _get_normalized_masked_log_probablities(vector, mask):
+    # We calculate normalized log probabilities in a numerically stable fashion, as done
+    # in https://github.com/rkadlec/asreader/blob/master/asreader/custombricks/softmax_mask_bricks.py
+    # TODO(mattg): a bunch of this logic can be simplified once pytorch-0.2 is out.
+    # torch.max(keepdim=True), for instance, simplifies things here.
+    # Here we get normalized log probabilities for enhanced numerical stability.
```
Comment (matt-gardner) [inline allennlp/common/tensor.py]: This line seems unnecessary now.

File: allennlp/common/tensor.py [hunk 2]
```diff
@@ -238,3 +260,60 @@ def weighted_sum(matrix: torch.Tensor, attention: torch.Tensor) -> torch.Tensor:
         matrix = matrix.expand(*expanded_size)
     intermediate = attention.unsqueeze(-1).expand_as(matrix) * matrix
     return intermediate.sum(dim=-2).squeeze(-2)
+
+
+def weighted_cross_entropy_with_logits(logits: torch.FloatTensor,
```
Comment (matt-gardner) [inline allennlp/common/tensor.py]: I really think this name should reflect that its intended use is for sequential targets.  It'd help to avoid confusion about when it should be used (like my initial confusion that I'd need this for BiDAF, because it handled masking).
